### ISSUE TITLE
Easier change optimizer/learning rate instead of reading state_dict from checkpoint

### ISSUE THREAD

**Issue by IncubatorShokuhou** [PyTorchLightning/pytorch-lightning#4044]
## 🚀 Feature
Easier change optimizer / learning rate without recoding

### Motivation

Usually when we want to change the optimizer from one to another(for example,using Adam to fast init training and turn to SGD at the end) in PyTorch, we may have to load state_dict of the model and change then change the optimizer in the code.

In PytorchLightning, model state_dict, state of all optimizers, state of all learning rate schedulers etc. are automately saved in the lightning checkpoint. It seems that we will still have to load state_dict of the model from the checkpoint. (Similar issues also happen when we want to change the learning rate). 

Is it possible that we are able to directly change the code and use a new optimizer/ new learning rate instead of recoding and reading the state_dict from checkpoint?

### Pitch

Directly change the code and then a  new optimizer/ new learning rate could be used.

**Comment by rohitgr7** [comment]
so you don't want to load the optimizer state from the checkpoint. Is this what you need?

**Comment by awaelchli** [comment]
Could you also post a code sample? It would help us better understand what you need.

**Comment by IncubatorShokuhou** [comment]
> so you don't want to load the optimizer state from the checkpoint. Is this what you need?

Yes. That is exactly what I need.

**Comment by IncubatorShokuhou** [comment]
> Could you also post a code sample? It would help us better understand what you need.

for example,
```python
import os
import torch
from torch.utils.data import Dataset
from pytorch_lightning import Trainer, LightningModule


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"loss": loss}

    def configure_optimizers(self):
        # optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        optimizer = optimizer = torch.optim.Adam(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]


def run_test():
    class TestModel(BoringModel):

        def on_train_epoch_start(self) -> None:
            print('override any method to prove your bug')

    # fake data
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    val_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    test_data = torch.utils.data.DataLoader(RandomDataset(32, 64))

    # model
    model = TestModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        max_epochs=100,
        weights_summary=None,
    )
    trainer.fit(model, train_data, val_data)
    # trainer.test(test_dataloaders=test_data)

if __name__ == '__main__':
    run_test()
```
when I finish trainning and want to switch
``` 
optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
```
to
```
optimizer = optimizer = torch.optim.Adam(self.layer.parameters(), lr=0.1)
```
, then error occours. If I want to use the weight of the trained model , I will have to do something like
```
state_dict = torch.load("./lightning_logs/version_0/checkpoints/epoch=99.ckpt")
model.load_state_dict(state_dict['model'])
optim = torch.optim.SGD(model.paramters(), new_learning_rate)

```
,which is too troublesome. So I am wondering if it is possible to do it more convenient.

**Comment by rohitgr7** [comment]
either you can set `save_weights_only=True` in `ModelCheckpoint` or pop the `optimizer_states` in `on_load_checkpoint` hook from your checkpoint and change the optimizer in `on_train_end` hook as [mentioned here](https://forums.pytorchlightning.ai/t/how-to-switch-from-optimizer-during-training/219/20). Although I am still confused about why you want to change the optimizer at the end of training.

**Comment by IncubatorShokuhou** [comment]
@rohitgr7 This idea is inspired by [Adabound](https://github.com/Luolc/AdaBound), in order to accelerate training at the beginning of the experiment.

**Comment by angshine** [comment]
I think it would be better allowing users to load `state_dict` from checkpoints with customization, e.g. sometimes I just want to load the model and optimizer states but not the scheduler. 

code ref: [training state loading](https://github.com/PyTorchLightning/pytorch-lightning/blob/3abfec896212ea85e45d6ac3ccb323ef242d16de/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L125)

**Comment by rohitgr7** [comment]
Then will it make sense to make it a warning rather than raising an error if `optimizer_states` or `lr_scheduler_states` are not found?
cc @PyTorchLightning/core-contributors

**Comment by rohitgr7** [comment]
here, should we make it a warning rather than raising an error if optimizer_states or lr_scheduler_states are not found while using resume_from_checkpoint?
cc @PyTorchLightning/core-contributors

**Comment by eepaul** [comment]
> either you can set `save_weights_only=True` in `ModelCheckpoint` or pop the `optimizer_states` in `on_load_checkpoint` hook from your checkpoint and change the optimizer in `on_train_end` hook as [mentioned here](https://forums.pytorchlightning.ai/t/how-to-switch-from-optimizer-during-training/219/20). Although I am still confused about why you want to change the optimizer at the end of training.

I have the same requirement.
I would like to use one learning rate without scheduler and start training. In the meantime, I observe metrics curves then decide at which epoch to decay the learning rate. Then I stop this training, load checkpoint at that epoch with a new decayed learning rate and start training again.

**Comment by JiamingSuen** [comment]
> I think it would be better allowing users to load `state_dict` from checkpoints with customization, e.g. sometimes I just want to load the model and optimizer states but not the scheduler.
> 
> code ref: [training state loading](https://github.com/PyTorchLightning/pytorch-lightning/blob/3abfec896212ea85e45d6ac3ccb323ef242d16de/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L125)

Save request here, please consider this @PyTorchLightning/core-contributors 
(why my team mentioning is not working?)

**Comment by 0x23** [comment]
I am also interested in this feature. Sometimes during training I see that the LR scheduler decreased the LR to early - so I would like to load the current checkpoint an continue with a diffrent lr - otherwise I would have to restart training wich is quite time consuming.
